File: src/PyTorch_CUDA_extension/src/lstm.c
#include <TH.h>
#include <THC.h>
#include <THCGeneral.h>
#include <time.h>

#include "lstm_kernel.h"

extern THCState* state;

int lstm_forward(
    THFloatTensor* h_data,
    THFloatTensor* x_data,
    THFloatTensor* c_data,
    THIntTensor* hiddenSize, 
    THIntTensor* miniBatch, 
    THIntTensor* seqLength, 
    THIntTensor* numLayers) {

	clock_t start,end;
	float e_time;
	start = clock();
	
    forward(state, h_data, x_data, c_data, 
            hiddenSize, miniBatch, seqLength, numLayers);

	end = clock();
	e_time = ((float)(end - start)) / CLOCKS_PER_SEC;
	printf("C wrapper time:\t%f\n", e_time);

    return 1;
}

File: src/PyTorch_CUDA_extension/lstm.py
import time
import torch
import _ext.lstm

class lstm(torch.autograd.Function):
    def __init__(self):
        super(lstm, self).__init__()

    def forward(self, h_data, x_data, c_data, hiddenSize, miniBatch, seqLength, numLayers):
        # if h_data.is_cuda == True and x_data.is_cuda == True and c_data.is_cuda == True:

        start_t = time.time()

        _ext.lstm.lstm_forward(h_data, x_data, c_data, hiddenSize, miniBatch, seqLength, numLayers)
        # else:
        #     raise NotImplementedError()

        elapsed_time = time.time() - start_t
        print("Function wrapper time:\t%f seconds"%(elapsed_time))

        # be forced to return a Tensor by autograd.Function
        return h_data
